examples/example.c

```diff
@@ -0,0 +1,248 @@
+        [MCE_DISPLAY_STATE_DIM] = "dim",
+        [MCE_DISPLAY_STATE_ON]  = "on",
+    };
+    return lut[state];
+}
+
+static const char *tklock_mode_repr(MCE_TKLOCK_MODE mode)
+{
+    static const char * const lut[] = {
+        [MCE_TKLOCK_MODE_LOCKED]            = "locked",
+        [MCE_TKLOCK_MODE_SILENT_LOCKED]     = "silent_locked",
+        [MCE_TKLOCK_MODE_LOCKED_DIM]        = "locked_dim",
+        [MCE_TKLOCK_MODE_LOCKED_DELAY]      = "locked_delay",
+        [MCE_TKLOCK_MODE_SILENT_LOCKED_DIM] = "silent_locked_dim",
+        [MCE_TKLOCK_MODE_UNLOCKED]          = "unlocked",
+        [MCE_TKLOCK_MODE_SILENT_UNLOCKED]   = "silent_unlocked",
+    };
+    return lut[mode];
+}
+
+static void battery_cb(MceBattery *battery, void *arg)
+{
+    (void)arg;
+    printf("battery: valid=%s level=%d status=%s\n",
+           bool_repr(battery->valid),
+           battery->level,
+           battery_status_repr(battery->status));
+}
+
+static void charger_cb(MceCharger *charger, void *arg)
+{
+    (void)arg;
+    printf("charger: valid=%s state=%s\n",
+           bool_repr(charger->valid),
+           charger_state_repr(charger->state));
+}
+
+static void display_cb(MceDisplay *display,void *arg)
+{
+    (void)arg;
+    printf("display: valid=%s state=%s\n",
+           bool_repr(display->valid),
+           display_state_repr(display->state));
+}
+
+static void tklock_cb(MceTklock *tklock, void *arg)
+{
+    (void)arg;
+    printf("tklock: valid=%s mode=%s locked=%s\n",
+           bool_repr(tklock->valid),
+           tklock_mode_repr(tklock->mode),
+           bool_repr(tklock->locked));
+}
+
+static void inactivity_cb(MceInactivity *inactivity, void *arg)
+{
+    (void)arg;
+    printf("inactivity: valid=%s status=%s\n",
+           bool_repr(inactivity->valid),
+           bool_repr(inactivity->status));
+}
+
+/* ========================================================================= *
+ * MAIN_ENTRY
+ * ========================================================================= */
+
+static gboolean quit_cb(gpointer aptr)
+{
+    printf("quit\n");
+    mainloop_quit();
+    guint *id_ptr = aptr;
+    *id_ptr = 0;
+    return G_SOURCE_REMOVE;
+}
+
+int
+main(int argc, char **argv)
+{
+    printf("startup\n");
+
+    int            exitcode       = EXIT_FAILURE;
+    gint           timeout_s      = 0;
+    guint          timeout_id     = 0;
+    guint          sigterm_id     = 0;
+    guint          sigint_id      = 0;
+    MceBattery    *mce_battery    = mce_battery_new();
+    MceCharger    *mce_charger    = mce_charger_new();
+    MceDisplay    *mce_display    = mce_display_new();
+    MceInactivity *mce_inactivity = mce_inactivity_new();
+    MceTklock     *mce_tklock     = mce_tklock_new();
+
+    /* Note: As handler callbacks do not use user data -> implicit signal
+     *       disconnect at final gobject unref is enough and there is no
+     *       need to keep track of added signal handler ids. */
```

Comment:
While this is true in this particular case (because global variables are being used instead of the context argument which IMO isn't a generally good practice), I think a sample should demonstrate a typical pattern - and a typical pattern would be to use the context parameter and require unregistering the handlers.

Reply:
Added explicit cleanup. And used args for something useful.

Side note: having xxx_new() functions return something that is actually a shared instance rather than something "new" is slightly confusing from assumed object lifetime point of view.

Reply:
Well, when refcounting is being used, no one really controls the object's lifetime anyway, what you do with the object shouldn't depend on how many times it's referenced and by whom. The only thing you really own is your reference.

Reply:
I guess we agree on necessity of cleaning up handlers. The delta is somewhere in (possibly misguided) expectations of mine / documentation realm. I mean if I know that I'm the only one holding a reference, then I know callbacks will be cleaned up implicitly at unref. Now I was slightly surprised by new() returning a potentially shared object which makes further cleanup necessary in order not to leave stale handlers behind in case somebody else is holding a ref too...